source/neuropod/bindings/java/src/main/native/com_uber_neuropod_Neuropod.cc

```diff
@@ -115,3 +117,71 @@ JNIEXPORT jstring JNICALL Java_com_uber_neuropod_Neuropod_nativeGetPlatform(JNIE
     }
     return nullptr;
 }
+
+JNIEXPORT jobject JNICALL Java_com_uber_neuropod_Neuropod_nativeGetInputs(JNIEnv *env, jclass, jlong handle)
+{
+    try
+    {
+        auto model     = reinterpret_cast<neuropod::Neuropod *>(handle);
+        auto inputSpec = model->get_inputs();
+        return toJavaTensorSpecList(env, inputSpec);
+    }
+    catch (const std::exception &e)
+    {
+        throwJavaException(env, e.what());
+    }
+    return nullptr;
+}
+
+JNIEXPORT jobject JNICALL Java_com_uber_neuropod_Neuropod_nativeGetOutputs(JNIEnv *env, jclass, jlong handle)
+{
+    try
+    {
+        auto model   = reinterpret_cast<neuropod::Neuropod *>(handle);
+        auto outputs = model->get_outputs();
+        return toJavaTensorSpecList(env, outputs);
+    }
+    catch (const std::exception &e)
+    {
+        throwJavaException(env, e.what());
+    }
+    return nullptr;
+}
+
+jobject toJavaTensorSpecList(JNIEnv *env, const std::vector<neuropod::TensorSpec> &specs)
+{
+    jobject ret = env->NewObject(java_util_ArrayList, java_util_ArrayList_, specs.size());
+    for (const auto &tensorSpec : specs)
+    {
+        auto    type = getFieldObject(env, com_uber_neuropod_TensorType, tensorTypeToString(tensorSpec.type).c_str());
+        jstring name = env->NewStringUTF(tensorSpec.name.c_str());
+        jobject dims = env->NewObject(java_util_ArrayList, java_util_ArrayList_, tensorSpec.dims.size());
+        for (const auto &dim : tensorSpec.dims)
+        {
+            // Dim is symbol
+            if (dim.value == -2)
+            {
+                jstring symbol = env->NewStringUTF(dim.symbol.c_str());
+                jobject javaDim =
+                    env->NewObject(com_uber_neuropod_Dimension, com_uber_neuropod_Dimension_symbol_, symbol);
+                env->CallBooleanMethod(dims, java_util_ArrayList_add, javaDim);
+                env->DeleteLocalRef(javaDim);
+                env->DeleteLocalRef(symbol);
+            }
+            else
+            {
+                jobject javaDim =
+                    env->NewObject(com_uber_neuropod_Dimension, com_uber_neuropod_Dimension_value_, dim.value);
+                env->CallBooleanMethod(dims, java_util_ArrayList_add, javaDim);
+                env->DeleteLocalRef(javaDim);
+            }
+        }
+        jobject javaTensorSpec =
+            env->NewObject(com_uber_neuropod_TensorSpec, com_uber_neuropod_TensorSpec_, name, type, dims);
+        env->CallBooleanMethod(ret, java_util_ArrayList_add, javaTensorSpec);
+        env->DeleteLocalRef(name);
+        env->DeleteLocalRef(dims);
+        env->DeleteLocalRef(type);
```

Comment:
We don't need to explicitly call `DeleteLocalRef` (here and in all the JNI code).

See [here](https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/design.html#wp1242) in the JNI specification.

Specificially,
> Local references are valid for the duration of a native method call, and are automatically freed after the native method returns.

> In most cases, the programmer should rely on the VM to free all local references after the native method returns. However, there are times when the programmer should explicitly free a local reference. 

There are some cases where we'd want to explicitly call `DeleteLocalRef` (e.g. large loop that creates objects, references to large objects, etc), but in most cases we should be fine without. See the spec linked above for more details.

I forgot to mention this on your previous PR. If you get a chance, can you put up another PR removing all the unnecessary `DeleteLocalRef`s? Otherwise, let's create an issue to clean this up.

Reply:
We need to delete here, because JNI only supports max to 16 local references, and in a loop it is likely beyond this limitation.

Reply:
Hmm I thought it was 512 local references, but from what I've read it's version specific.

That said, I missed that these deletes are in the loop (I thought they were at the end if the function). Let's leave them here, but in other spots where we're not in a loop, it'll probably simplify code to remove `DeleteLocalRef`